zerver/worker/queue_processors.py

```diff
@@ -115,26 +115,84 @@ def on_failure(event: Dict[str, Any]) -> None:
+    CONSUME_ITERATIONS_BEFORE_UPDATE_STATS_NUM = 50
 
     def __init__(self) -> None:
         self.q = None  # type: SimpleQueueClient
         if self.queue_name is None:
             raise WorkerDeclarationException("Queue worker declared without queue_name")
 
+        self.initialize_statistics()
+
+    def initialize_statistics(self) -> None:
+        self.queue_last_emptied_timestamp = time.time()
+        self.consumed_since_last_emptied = 0
+        self.recent_consume_times = deque(maxlen=50)  # type: MutableSequence[Tuple[int, float]]
+        self.consume_interation_counter = 0
+
+        self.update_statistics(0)
+
+    def update_statistics(self, remaining_queue_size: int) -> None:
+        total_seconds = sum([seconds for _, seconds in self.recent_consume_times])
+        total_events = sum([events_number for events_number, _ in self.recent_consume_times])
+        if total_events == 0:
+            recent_average_consume_time = None
+        else:
+            recent_average_consume_time = total_seconds / total_events
+        stats_dict = dict(
+            update_time=time.time(),
+            recent_average_consume_time=recent_average_consume_time,
+            current_queue_size=remaining_queue_size,
+            queue_last_emptied_timestamp=self.queue_last_emptied_timestamp,
+            consumed_since_last_emptied=self.consumed_since_last_emptied,
+        )
+
+        os.makedirs(settings.QUEUE_STATS_DIR, exist_ok=True)
+
+        fname = '%s.stats' % (self.queue_name,)
+        fn = os.path.join(settings.QUEUE_STATS_DIR, fname)
+        with lockfile(fn + '.lock'):
+            tmp_fn = fn + '.tmp'
+            with open(tmp_fn, 'w') as f:
+                ujson.dump(stats_dict, f, indent=2)
+            os.rename(tmp_fn, fn)
+
     @abstractmethod
     def consume(self, data: Dict[str, Any]) -> None:
         pass
 
     def do_consume(self, consume_func: Callable[[List[Dict[str, Any]]], None],
                    events: List[Dict[str, Any]]) -> None:
         try:
+            time_start = time.time()
             consume_func(events)
+            consume_time_seconds = time.time() - time_start  # type: Optional[float]
+            self.consumed_since_last_emptied += len(events)
         except Exception:
             self._handle_consume_exception(events)
+            consume_time_seconds = None
         finally:
             flush_per_request_caches()
             reset_queries()
 
+            if consume_time_seconds is not None:
+                self.recent_consume_times.append((len(events), consume_time_seconds))
+
+            if self.q is not None:
+                remaining_queue_size = self.q.queue_size()
+            else:
+                remaining_queue_size = 0
+
+            if remaining_queue_size == 0:
+                self.queue_last_emptied_timestamp = time.time()
+                self.consumed_since_last_emptied = 0
+
+            self.consume_interation_counter += 1
```

Comment:
What do you think about making this `len(events)` instead, rather than 1?  I noticed it took a few minutes for `user_activity.stats` to first get written to on chat.zulip.org, and was thinking this might be because we're not treating LoopQueueProcessingWorkers properly in deciding how often to log stats data.